File: ios/SuwappuApp/SuwappuApp/Models/Swap.swift
//
//  Swap.swift
//  SuwappuApp
//
//  Created on [Date]
//

import Foundation

enum SwapStatus: String, Codable {
    case pending = "pending"
    case quoteReceived = "quote_received"
    case awaitingApproval = "awaiting_approval"
    case approved = "approved"
    case executing = "executing"
    case submitted = "submitted"
    case confirming = "confirming"
    case completed = "completed"
    case failed = "failed"
    case cancelled = "cancelled"
    
    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .quoteReceived: return "Quote Received"
        case .awaitingApproval: return "Awaiting Approval"
        case .approved: return "Approved"
        case .executing: return "Executing"
        case .submitted: return "Submitted"
        case .confirming: return "Confirming"
        case .completed: return "Completed"
        case .failed: return "Failed"
        case .cancelled: return "Cancelled"
        }
    }
    
    var isCompleted: Bool {
        return self == .completed
    }
    
    var isFailed: Bool {
        return self == .failed || self == .cancelled
    }
    
    var isPending: Bool {
        return [.pending, .quoteReceived, .awaitingApproval, .approved, .executing, .submitted, .confirming].contains(self)
    }
}

struct SwapTransaction: Identifiable, Codable {
    let id: Int
    let userId: Int
    let fromChain: String
    let fromToken: String
    let fromAmount: String
    let fromAmountUSD: Double?
    let toChain: String
    let toToken: String
    let toAmount: String?
    let toAmountUSD: Double?
    let status: SwapStatus
    let txHash: String?
    let bridgeTxHash: String?
    let destinationTxHash: String?
    let routeProvider: String?
    let gasFee: Double?
    let bridgeFee: Double?
    let slippage: Int
    let createdAt: Date
    let updatedAt: Date
    let completedAt: Date?
    let errorMessage: String?
    
    var isCrossChain: Bool {
        return fromChain != toChain
    }
}

struct SwapQuote: Codable {
    let provider: String // "lifi", "jupiter", "layerzero", "ccip"
    let fromChain: String
    let toChain: String
    let fromToken: String
    let toToken: String
    let fromAmount: String
    let fromAmountHuman: Double
    let toAmount: String
    let toAmountHuman: Double
    let toAmountMin: String
    let gasCostUSD: Double
    let feeCostUSD: Double
    let totalCostUSD: Double
    let estimatedTime: Int // seconds
    let priceImpact: Double
    let exchangeRate: Double
    let expiresIn: Int // seconds
    let timestamp: Date
}

struct SwapQuoteRequest: Codable {
    let fromChain: String
    let toChain: String
    let fromToken: String
    let toToken: String
    let amount: Double
    let fromAddress: String
    let toAddress: String?
    let slippage: Double // percentage
}

struct ExecuteSwapRequest: Codable {
    let quote: SwapQuote
    let walletId: Int
}

struct SwapHistoryResponse: Codable {
    let swaps: [SwapTransaction]
    let total: Int
    let page: Int
    let pageSize: Int
}

File: packages/shared/src/types/swap.ts
/**
 * Swap-related types for Suwappu
 * Extracted from webapp/src/types/swap.ts for cross-platform use
 */

export interface SwapToken {
  symbol: string
  name: string
  address: string
  chain: string
  decimals: number
  logoUrl?: string
  balance?: string
  balanceUsd?: number
}

export interface SwapQuoteRequest {
  fromToken: string
  toToken: string
  fromChain: string
  toChain: string
  amount: string
  fromDecimals: number
  slippage?: number
}

export interface SwapQuote {
  id: string
  fromToken: SwapToken
  toToken: SwapToken
  fromAmount: string
  toAmount: string
  fromAmountUsd: number
  toAmountUsd: number
  exchangeRate: number
  priceImpact: number
  estimatedGas: string
  gasUsd: number
  route: string
  expiresAt: string
  minReceived: string
  slippage: number
  estimatedDuration?: number
}

export interface SwapExecuteRequest {
  quoteId: string
}

export interface SwapExecuteResult {
  success: boolean
  swapId: number
  status: 'signed' | 'submitted' | 'completed' | 'failed'
  txHash?: string
  explorerUrl?: string
  swap: {
    fromChain: string
    toChain: string
    fromToken: string
    toToken: string
    fromAmount: string
    expectedToAmount: string
  }
}

export interface SwapStatusResponse {
  id: number
  status: 'pending' | 'signed' | 'submitted' | 'completed' | 'failed'
  fromChain: string
  toChain: string
  fromToken: string
  toToken: string
  fromAmount: string
  toAmount: string | null
  txHash: string | null
  bridgeTxHash: string | null
  destinationTxHash: string | null
  errorMessage: string | null
  createdAt: string
  completedAt: string | null
}
